dll/win32/iphlpapi/icmp.c

```diff
@@ -1,1129 +1,575 @@
-                    TRACE("Got address %s\n", inet_ntoa(addr));
-                }
-            }
-
-            /* Interpret ttl */
-            if ((ptr = strstr(buf, "ttl=")))
-            {
-                int val;
-                if (sscanf(ptr + 4, "%u", &val) >= 1)
-                {
-                    reply->Options.Ttl = val;
-                    TRACE("Got ttl %u\n", val);
-                }
-            }
-
-            /* Interpret time */
-            if ((ptr = strstr(buf, "time=")))
-            {
-                float val;
-                if (sscanf(ptr + 5, "%f", &val) >= 1)
-                {
-                    reply->RoundTripTime = (unsigned int)(val + 0.5);
-                    TRACE("Got rtt = %u\n", reply->RoundTripTime);
-                }
-            }
-
-            memmove(buf, eol + 1, pos - (eol + 1 - buf));
-            pos -= (eol + 1 - buf);
+            CopyMemory(
+                (PBYTE)InputBuffer + sizeof(ICMP_ECHO_REQUEST),
+                RequestOptions->OptionsData,
+                Request->OptionsSize);
         }
     }
-    while (res > 0);
-    close(pipe_out[0]);
 
-    /* reap the child process */
-    do
+    if (RequestSize > 0)
     {
-        wpid = waitpid(pid, &status, 0);
+        Request->DataSize = RequestSize;
+        CopyMemory((PBYTE)InputBuffer + Request->DataOffset, RequestData, RequestSize);
     }
-    while (wpid < 0 && errno == EINTR);
 
-    /* fill out remaining struct fields */
-    if (wpid >= 0 && WIFEXITED(status) && WEXITSTATUS(status) == 0)
+    if (Event == NULL && ApcRoutine == NULL)
     {
-        if (ReplySize < RequestSize + sizeof(*reply))
-        {
-            reply->Status   = IP_BUF_TOO_SMALL;
-            reply->DataSize = 0;
-            reply->Data     = NULL;
-        }
-        else
-        {
-            reply->Status   = 0;
-            reply->DataSize = RequestSize;
-            reply->Data     = (char *)reply + sizeof(*reply);
-            memcpy(reply->Data, RequestData, RequestSize);
-        }
-        return 1;
+        hEvent = CreateEventW(NULL, FALSE, FALSE, NULL);
     }
-
-    SetLastError(IP_REQ_TIMED_OUT);
-    return 0;
-#else
-    ERR("no fork support on this platform\n");
-    SetLastError(ERROR_NOT_SUPPORTED);
-    return 0;
-#endif
-}
-#else // __REACTOS__
-BOOL
-GetIPv4ByIndex(
-    _In_  DWORD              Index,
-    _Out_ IPAddr *           Address
-)
-{
-    PMIB_IPADDRTABLE pIpAddrTable;
-    ULONG dwSize = 0;
-    BOOL result = FALSE;
-
-    if (GetIpAddrTable(NULL, &dwSize, FALSE) != ERROR_INSUFFICIENT_BUFFER)
+    else
+    {
+        hEvent = Event;
+    }
+
+    Status = NtDeviceIoControlFile(
+        IcmpHandle,
+        hEvent,
+        ApcRoutine,
+        ApcContext,
+        &IoStatusBlock,
+        IOCTL_ICMP_ECHO_REQUEST,
+        InputBuffer,
+        ReplySize,
+        ReplyBuffer,
+        ReplySize);         // TODO: Determine how Windows calculates OutputBufferLength.
```

Comment:
I don't understand this comment.

Reply:
IIRC, it it was something I saw when inspecting the ioctl behavior, where Microsoft's implementation would pass different buffer sizes to the the driver. I never did find out why, but using a reply buffer of the same size as the input buffer worked on Windows, so I don't think it's significant.

@Extravert-ir should ensure these call sites conform to his implementation and can remove the TODOs.